### PR TITLE
Task 2 2 1

### PR DESCRIPTION
**Pull request by benmasud**

### CODE REVIEW

File: .gradle/buildOutputCleanup/cache.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+#Tue Mar 14 09:13:05 NOVT 2023
+gradle.version=7.5.1
```
Comment (molo4evan) [inline .gradle/buildOutputCleanup/cache.properties:2]: You have added .idea and gradle files before gitignore, as i see

You should remove it from your index, see `git rm --cached`

File: Task_1_1_2/Task_1_2_2_/src/main/java/r/masud/fit/nsu/ru/Main.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package r.masud.fit.nsu.ru;
+
+public class Main {
```
Comment (molo4evan) [inline Task_1_1_2/Task_1_2_2_/src/main/java/r/masud/fit/nsu/ru/Main.java:3]: Remove excess files from PR

File: Task_2_1_1/build/classes/java/main/META-INF/BenchmarkList [hunk 3]
```diff
@@ -0,0 +1,3 @@
+JMH S 31 r.masud.ParallelStreamBenchmark S 70 r.masud.jmh_generated.ParallelStreamBenchmark_containsNonPrime_jmhTest S 16 containsNonPrime S 11 AverageTime E A 1 1 1 E E E E E E E E E E E E E M 1 4 size 3 16 xAAMAADAwAA===== 16 xAAMAADAwAAMAA== 16 xAAMAADAwAAMAADA U 12 MILLISECONDS E E
```
Comment (molo4evan) [inline Task_2_1_1/build/classes/java/main/META-INF/BenchmarkList:1]: build directory should be also removed from index

You can ask your groupmates for help with gitignore

File: Task_2_1_1/src/main/java/r/masud/ParallelStreamBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package r.masud;
+
+import org.openjdk.jmh.annotations.*;
+import java.util.Arrays;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+@BenchmarkMode(Mode.AverageTime)
```
Comment (molo4evan) [inline Task_2_1_1/src/main/java/r/masud/ParallelStreamBenchmark.java:8]: And also 2.1.1 should not be in PR by 2.2.1

You should keep labs in separate branches and branch them all from main, not from each other

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Main {
```
Comment (molo4evan) [inline Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java:14]: image

In this PR you should only have Task 2_2_1 directory

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    public static void main(String[] args) {
+        // Read JSON input file
+        JsonParser parser = new JsonParser();
+        try (FileReader reader = new FileReader("C:\\Users\\Asus\\IdeaProjects\\oop\\Task_2_2_1\\MeinePizza.json")) {
```
Comment (molo4evan) [inline Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java:18]: What if I want to download and run your code? Will it work?

https://www.geeksforgeeks.org/loading-resources-from-classpath-in-java-with-example/

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            JsonElement root = parser.parse(reader);
+            JsonArray ordersJsonArray = root.getAsJsonArray();
+
+            // Convert JSON to objects
+            Gson gson = new GsonBuilder().create();
```
Comment (molo4evan) [inline Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java:23]: You should load from JSON the configuration of pizzeria (baker and courier paramters and count, warehouse capacity, etc.), not orders

For orders the best option is to create Clients, which will generate orders on the fly in separate threads.

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+            // Start bakery and courier threads
+            Bakery bakery = new Bakery(pizzaQueue, warehouse);
+            CourierThread courierThread = new CourierThread(courier, warehouse);
+            bakery.start();
+            courierThread.start();
+
+            // Wait for threads to finish
+            try {
+                bakery.join();
```
Comment (molo4evan) [inline Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java:54]: By default your pizzeria should support infinite mode of work and provide methods such as "start", "stop", "exit"

Also you should implement "graceful" and "force" shutdown:
1) You just stop clients, but all orders already posted will be finished
2) You completely finish the work of your pizzeria, even if you have not done orders

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Warehouse.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
+package r.nsu.fit.masud;
+
+import java.util.LinkedList;
+import java.util.Queue;
+
+public class Warehouse {
```
Comment (molo4evan) [inline Task_2_2_1/src/main/java/r/nsu/fit/masud/Warehouse.java:6]: Your warehouse should provide safe methods for different thread could work with it

See Java synchronized, wait, notify, notifyAll

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                Order order = gson.fromJson(element, Order.class);
+                orders.add(order);
+            }
+
+            // Create objects
+            PizzaQueue pizzaQueue = new PizzaQueue();
+            Warehouse warehouse = new Warehouse(100);
+            Courier courier = new Courier(50);
+
+            for (Order order : orders) {
```
Comment (molo4evan) [inline Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java:35]: The main logic of pizzeria should be incapsulated to separate class

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Main {
+    public static void main(String[] args) {
```
Comment (molo4evan) [inline Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java:15]: Provide the tests:
Warehouse work
Graceful shutdown
Force shutdown

File: .gradle/buildOutputCleanup/cache.properties (hunk 1 above)
Comment (rafi2002) [reply .gradle/buildOutputCleanup/cache.properties:2]: fixed

File: Task_1_1_2/Task_1_2_2_/src/main/java/r/masud/fit/nsu/ru/Main.java (hunk 2 above)
Comment (rafi2002) [reply Task_1_1_2/Task_1_2_2_/src/main/java/r/masud/fit/nsu/ru/Main.java:3]: ah, understand .

File: Task_2_1_1/build/classes/java/main/META-INF/BenchmarkList (hunk 3 above)
Comment (rafi2002) [reply Task_2_1_1/build/classes/java/main/META-INF/BenchmarkList:1]: okay )

File: Task_2_1_1/src/main/java/r/masud/ParallelStreamBenchmark.java (hunk 4 above)
Comment (rafi2002) [reply Task_2_1_1/src/main/java/r/masud/ParallelStreamBenchmark.java:8]: yes, but I had a minor problem so I had to brunch them from diffrent, root. I'II fix it with the next task

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java (hunk 5 above)
Comment (rafi2002) [reply Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java:14]: understood

File: Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java (hunk 6 above)
Comment (rafi2002) [reply Task_2_2_1/src/main/java/r/nsu/fit/masud/Main.java:18]: yes, it will. I ran it sevral time before PR.

**Pull request closed without merging**